homeassistant/components/simplisafe/config_flow.py

```diff
@@ -98,146 +67,75 @@ def async_get_options_flow(
-            # in that case, we reauth them by making them go through the user flow:
-            return await self.async_step_user()
-
-        self._username = entry_data[CONF_USERNAME]
-        return await self.async_step_reauth_confirm()
-
-    async def _async_get_email_2fa(self) -> None:
-        """Define a task to wait for email-based 2FA."""
-        assert self._simplisafe
-
-        try:
-            async with async_timeout.timeout(DEFAULT_EMAIL_2FA_TIMEOUT):
-                while True:
-                    try:
-                        await self._simplisafe.async_verify_2fa_email()
-                    except Verify2FAPending:
-                        LOGGER.info("Email-based 2FA pending; trying again")
-                        await asyncio.sleep(DEFAULT_EMAIL_2FA_SLEEP)
-                    else:
-                        break
-        finally:
-            self.hass.async_create_task(
-                self.hass.config_entries.flow.async_configure(flow_id=self.flow_id)
-            )
-
-    async def async_step_email_2fa(
+    async def async_step_input_auth_code(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
-        """Handle email-based two-factor authentication."""
-        if not self._email_2fa_task:
-            self._email_2fa_task = self.hass.async_create_task(
-                self._async_get_email_2fa()
-            )
-            return self.async_show_progress(
-                step_id="email_2fa", progress_action="email_2fa"
+        """Handle the input of a SimpliSafe OAuth authorization code."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="input_auth_code", data_schema=STEP_INPUT_AUTH_CODE_SCHEMA
             )
 
-        try:
-            await self._email_2fa_task
-        except asyncio.TimeoutError:
-            return self.async_show_progress_done(next_step_id="email_2fa_error")
-        return self.async_show_progress_done(next_step_id="finish")
+        assert self._oauth_values
 
-    async def async_step_email_2fa_error(
-        self, user_input: dict[str, Any] | None = None
-    ) -> FlowResult:
-        """Handle an error during email-based two-factor authentication."""
-        return self.async_abort(reason="email_2fa_timed_out")
+        self._errors = {}
+        session = aiohttp_client.async_get_clientsession(self.hass)
 
-    async def async_step_finish(
-        self, user_input: dict[str, Any] | None = None
-    ) -> FlowResult:
-        """Handle the final step."""
-        assert self._simplisafe
-        assert self._username
+        try:
+            simplisafe = await API.async_from_auth(
+                user_input[CONF_AUTH_CODE],
+                self._oauth_values.code_verifier,
+                session=session,
+            )
+        except InvalidCredentialsError:
+            self._errors = {"base": "invalid_auth"}
```

Comment:
Errors are not meant to be shared with other steps. Errors are form specific and should only be local within the step where they originate from.